src/cpp/client/client_context.cc

```diff
@@ -72,6 +72,22 @@ ClientContext::~ClientContext() {
   g_client_callbacks->Destructor(this);
 }
 
+void ClientContext::set_credentials(
+    const std::shared_ptr<grpc_impl::CallCredentials>& creds) {
+  creds_ = creds;
+  // If call_ is set, we have already created the call, and set the call
+  // credentials. This should only be done before we have started the batch
+  // for sending initial metadata.
+  if (creds_ != nullptr && call_ != nullptr) {
+    if (!creds_->ApplyToCall(call_)) {
+      SendCancelToInterceptors();
+      grpc_call_cancel_with_status(call_, GRPC_STATUS_CANCELLED,
+                                   "Failed to set credentials to rpc.",
+                                   nullptr);
```

Comment:
Should the cancellation also trigger if the creds_ has already been set? Although I'm not normally into protecting against API misuse, we already have the handling in place here for other reasons. This is a nit, and I'm ok with the answer being no so long as the API comment above makes it clear that it can only be called once.

Reply:
I think we should be allowing set_credentials to be called multiple times, especially if it is happening before initial metadata is sent.

Reply:
Added a test for that